### DISCUSSION TITLE
SQL on FHIR v2

### DISCUSSION THREAD

**Discussion by jdjkelly** [medplum/medplum#4661] [Ideas]
At DevDays 2024 SQL on FHIR v2 was a big topic.

* There are multiple independent implementations in production now
  * Google supports View Definitions in their [FHIR Views](https://developers.google.com/open-health-stack/fhir-analytics/fhir-views) project
  * NCQA has written a C++ implementation (incl., necessarily, a FHIR Path implementation) that goes from NDJSON to Parquet (which they later load back into Postgres)
  * Aidbox has an implementation that takes View Definitions and compiles them into Postgres Views somehow (they have a pg extension that can do FHIR Path stuff) - they also have a nice View Definition UI builder
  * Pathling has an implementation (not sure if there's a public link)
* Community is aiming for a tagged release this summer

Previously there was some discussion on [Discord about Medplum](https://discord.com/channels/905144809105260605/1230173071655178281/1230173071655178281) support. @rahul1 last mentioned:

> We're definitely open to it! 
> We've always wanted to support some kind of sql-like syntax for data engineers, and it seems like this spec is starting to gain adoption
> The main implementation design point will be around performance and multi-tenancy - how do we implement "view definitions" in a scalable way?

I think that an "offline" View Runner to start could be pretty productive:

* A package that implements the Validate and Process steps given a ViewDefinition and a Resource as inputs
* This package could be run as a CLI tool, with an input format of an NDJSON file (some result of $export operation)
* Initially, the output format of the View Runner could be CSV (we're starting super basic)

This would be a really quick implementation and tool to run small experiments with - plus having another independent View Runner implementation will increase the likelihood of a specification release.

**Comment by codyebberson** [comment]
PR opened to add the Medplum implementation to the official registry: https://github.com/FHIR/sql-on-fhir-v2/pull/250

**Comment by codyebberson** [comment]
Now listed on the implementations page: https://fhir.github.io/sql-on-fhir-v2/#impls
